util/ctxkey/key.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// ctxkey provides type-safe key-value pairs for use with [context.Context].
+//
+// Example usage:
+//
+//	// Create a context key.
+//	var DeleteCooldownKey = ctxkey.New("logz.DeleteCooldown", 5*time.Second)
+//
+//	// Store a context value.
+//	ctx = logz.DeleteCooldownKey.WithValue(ctx, 10*time.Second)
+//
+//	// Load a context value.
+//	cooldown := logz.DeleteCooldownKey.Value(ctx)
+//	... // use cooldown of type time.Duration
+//
+// This is inspired by https://go.dev/issue/49189.
+package ctxkey
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+)
+
+// Key is a generic key type associated with a specific value type.
+// The Value type must not be an interface type.
+type Key[Value any] struct {
+	name   *string
```

Comment:
Why is name a pointer? Is the nil value symbolic?

Reply:
For use with `context.WithValue`. The key we provide it must be universally unique so a `string` itself is insufficient. On the other hand, a pointer we allocate ourselves must have that property.

These allocations aren't costly since they occur once in the creation of the key, which are almost always statically created.

Reply:
Added a comment explaining this nuance.